utils/pattern/src/parser/mod.rs

```diff
@@ -0,0 +1,315 @@
+/// [`TryFrom<&str>`][`TryFrom`] gets implemented on all types that support [`FromStr`].
+///
+/// [`TR35 2.6.1]: https://unicode.org/reports/tr35/tr35-dates.html#dateTimeFormat
+/// [`RFC 2924`]: https://github.com/rust-lang/rfcs/pull/2924
+/// [`Item`]: std::iter::Iterator::Item
+/// [`TryFrom`]: std::convert::TryFrom
+pub struct Parser<'p> {
+    input: &'p str,
+    len: usize,
+
+    start_idx: usize,
+    idx: usize,
+
+    state: ParserState,
+}
+
+impl<'p> Parser<'p> {
+    /// Creates a new `Parser`.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::Parser;
+    /// let mut parser = Parser::new("{0}, {1}");
+    /// ```
+    pub fn new(input: &'p str) -> Self {
+        Self {
+            input,
+            len: input.len(),
+
+            start_idx: 0,
+            idx: 0,
+
+            state: ParserState::default(),
+        }
+    }
+
+    /// An iterator method that advances the iterator and returns the result of an attempt to parse
+    /// the next token.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::{Parser, PatternToken};
+    ///
+    /// let mut parser = Parser::new("{0}, {1}");
+    ///
+    /// // A call to try_next() returns the next value…
+    /// assert_eq!(Ok(Some(PatternToken::Placeholder(0))), parser.try_next());
+    /// assert_eq!(Ok(Some(PatternToken::Literal { content: ", ", quoted: false})), parser.try_next::<usize>());
+    /// assert_eq!(Ok(Some(PatternToken::Placeholder(1))), parser.try_next());
+    ///
+    /// // … and then None once it's over.
+    /// assert_eq!(Ok(None), parser.try_next::<usize>());
+    /// ```
+    pub fn try_next<P>(
+        &mut self,
+    ) -> Result<Option<PatternToken<'p, P>>, ParserError<<P as FromStr>::Err>>
+    where
+        P: FromStr,
+        P::Err: Debug,
+    {
+        while let Some(b) = self.input.as_bytes().get(self.idx) {
+            match self.state {
+                ParserState::Placeholder if *b == b'}' => {
+                    let range = advance_state!(self, self.idx, ParserState::Default);
+                    return self.input[range]
+                        .parse()
+                        .map(|ret| Some(PatternToken::Placeholder(ret)))
+                        .map_err(ParserError::InvalidPlaceholder);
+                }
+                ParserState::QuotedLiteral if *b == b'\'' => {
+                    handle_literal!(self, true, ParserState::Default)
+                }
+                ParserState::Default if *b == b'{' => {
+                    handle_literal!(self, false, ParserState::Placeholder)
+                }
+                ParserState::Default if *b == b'\'' => {
+                    handle_literal!(self, false, ParserState::QuotedLiteral)
+                }
+                ParserState::Default if b.is_ascii_alphabetic() => {
+                    return Err(ParserError::IllegalCharacter(*b as char));
+                }
```

Comment:
Issue: CLDR is inconsistent over when quoted literals are required.  I assume that when you wrote this, you were largely looking at the glue pattern type data, like `"{0} 'at' {1}"`:

https://unicode.org/reports/tr35/tr35-dates.html#dateTimeFormats

However, in other contexts, CLDR doesn't require quoted literals.  For example, in the same spec, scroll to the relative time formatting section and you will find `"in {0} days"`:

https://unicode.org/reports/tr35/tr35-dates.html#Calendar_Fields

I guess I would just allow unquoted ASCII characters and treat them as any other literal.

Reply:
I can make the parser accept a boolean flag that specifies how ASCII letters should be treated.

Reply:
Done